### PR TITLE
webpages to new/edit/list station

### PR DESCRIPTION
**Pull request by jeffwang0516**
1. finish pages of station management
2. finish handling form datas from the pages
3. tested on nginx+uwsgi, the image_url in station_list & get_all_stations can work
4. comments added
5. templates station_new.html & station_edit.html are with small differences on the image upload section
p.s The original add_station is changed to station_new (including templates and view)

### CODE REVIEW

File: smart_campus/app/templates/app/station_edit.html [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    <!-- Beacon Information -->
+    <div class="row">
+      <div class="col-md-5">
+        <div class="col-md-12">
+
+          <!-- Select linking Beacon from Map -->
+          <label class="control-label col" for="selectbasic">Beacon選擇(請在地圖點選)</label> 
+          <select id="selectbasic" name="beacon" class="custom-select" required="">
+            <option value>請在地圖點選...</option>
+            {% if beacons %}
```
Comment (Lee-W) [inline smart_campus/app/templates/app/station_edit.html]: It seems to me this `if` block is the same as line 34~38.
Please rewrite it as another tempalte.

File: smart_campus/app/templates/app/station_new.html [hunk 2]
```diff
@@ -5,51 +5,62 @@
...
       </div>
+
+      <!-- Category Select -->
       <div class="col-md-4">
         <label class="control-label col-md-12" for="select_category">類別</label> 
-        <select id="select_category" name="category" class="custom-select" value="{{form.cleaned_data.category}}" required="">
-          <option value selected>請選擇類別...</option>
+        <select id="select_category" name="category" class="custom-select" required="">
+          <option value>請選擇類別...</option>
           {% if categories %}
```
Comment (Lee-W) [inline smart_campus/app/templates/app/station_new.html]: Use the rewrote template I mentioned in `station_edit.html`.

File: smart_campus/app/templates/app/station_new.html [hunk 3]
```diff
@@ -59,44 +70,54 @@
...
     <div class="row">
       <div class="col-md-5">
         <div class="col-md-12">
+
+          <!-- Select linking Beacon from Map -->
           <label class="control-label col" for="selectbasic">Beacon選擇(請在地圖點選)</label> 
           <select id="selectbasic" name="beacon" class="custom-select" required="">
-            <option value selected>請在地圖點選...</option>
+            <option value>請在地圖點選...</option>
             {% if beacons %}
```
Comment (Lee-W) [inline smart_campus/app/templates/app/station_new.html]: Use the rewrote template I mentioned in station_edit.html.

File: smart_campus/app/views.py [hunk 4]
```diff
@@ -113,14 +119,145 @@ def index(request):
...
+        stations = Station.objects.filter(owner_group=request.user.group)
+
+    station_data = [
+        {
+            'id': station.id,
+            'name': station.name,
+            'image_url': (
+                '' if not station.stationimage_set.exists()
+                else
+                '/{0}'.format(station.stationimage_set.get(is_primary=True).image.url)
```
Comment (Lee-W) [inline smart_campus/app/views.py]: Is it possible to write an `image_url` property in Django model?

File: smart_campus/app/views.py [hunk 5]
```diff
@@ -113,14 +119,145 @@ def index(request):
...
+            if request.POST.get('img_changed') == 'true':
+                # Clear old images
+                for img in StationImage.objects.filter(station=station):
+                    # Clear files on the disk
+                    # Delete model instance won't delete them
+                    if os.path.isfile(img.image.path):
+                        os.remove(img.image.path)
+                    img.delete()
+
+                # Add new images (Max: 4 imgs)
```
Comment (Lee-W) [inline smart_campus/app/views.py]: Is there some reason you limit the number to 4?
If so, could you extract this as a configurable variable?

File: smart_campus/app/views.py [hunk 6]
```diff
@@ -113,14 +119,145 @@ def index(request):
...
+                for i in range(1, 5):
+                    if data['img{0}'.format(i)]:
```
Comment (Lee-W) [inline smart_campus/app/views.py]: `'img{0}'.format(i)` is used more multiple times.
Save it in a variable.

File: smart_campus/smart_campus/urls.py [hunk 7]
```diff
@@ -25,6 +25,7 @@
     url(r'^login/$', app.views.login_page, name='Login Page'),
     url(r'^logout/$', app.views.logout_page, name='Logout Page'),
     url(r'^$', app.views.index, name='index'),
-    url(r'^station_list/$', app.views.station_list_page, name='Station List Page'),
-    url(r'^add_station/$', app.views.add_station_page, name='Add Station Page'),
+    url(r'^stations/$', app.views.station_list_page, name='Station List Page'),
+    url(r'^stations/new/$', app.views.station_new_page, name='Add Station Page'),
+    url(r'^stations/(?P<pk>\d+)/$', app.views.station_edit_page, name='Edit Station Page'),
```
Comment (Lee-W) [inline smart_campus/smart_campus/urls.py]: Maybe a URL `^stations/(?P<pk>\d+)/edit/$` would be better.

File: smart_campus/app/views.py [hunk 8]
```diff
@@ -142,10 +279,10 @@ def get_all_stations(request):
...
             'location': station.location.get_coords(),
-            'image': [{'image_url': img.image.url, 'primary': img.is_primary}
+            'image': [{'image_url': 'http://{0}/{1}'.format(request.get_host(), img.image.url), 'primary': img.is_primary}
                       for img in station.stationimage_set.all()]
         }
         for station in Station.objects.all()
     ]
 
-    return JsonResponse({'status': 'true', 'message': 'Success', 'data': data})
+    return JsonResponse({'status': 'true', 'message': 'Success', 'data': data}, json_dumps_params={'ensure_ascii': False}, content_type='application/json; charset=utf-8')
```
Comment (Lee-W) [inline smart_campus/app/views.py]: Make sure all other `JsonResponse`s that would contain Chinese are set as this one.

File: smart_campus/app/views.py [hunk 9]
```diff
@@ -113,14 +119,145 @@ def index(request):
...
+                    StationImage.objects.create(
+                        station=station,
+                        image=data['img{0}'.format(img_num)],
+                        is_primary=True if img_num == data['main_img_num'] else False
+                    )
+            return HttpResponseRedirect('/stations/')
+    else:
+        form = StationForm()
+
+    if request.user.can(Permission.ADMIN):
```
Comment (Lee-W) [inline smart_campus/app/views.py]: I suggest you rewrite as follow.

```python
if request.user.can(Permission.EDIT):
    beacon_set = Beacon.objects.filter(owner_group=request.user.group)
elif request.user.can(Permission.ADMIN):
    beacon_set = Beacon.objects.all()
else:
    some error
```

File: smart_campus/app/static/app/js/station_new.js [hunk 10]
```diff
@@ -5,7 +5,7 @@ function initMap() {
     center: {lat: 22.997496, lng: 120.216508},
     zoom: 14
   });
-  //Add marker of beacons
+  // Add marker of beacons
   for(var i=0; i < location_arr.length ;i++){
```
[the remark this answers is not in the document]
Comment (jeffwang0516) [reply smart_campus/app/static/app/js/station_new.js]: yes

File: smart_campus/app/templates/app/station_edit.html [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    </div>
+    <br>
+
+    <!-- Choose Primary photo : HIDDEN-->
+    <div class="row" id="img_choose_sec" style="display: none;">
+      <div class="col-md-8" id='radios-img-select'> 
+        <label class="col-md-3 control-label" >選擇主要照片</label>
+        <label class="radio-inline" for="radios-1">
+          <input type="radio" name="main_img_num" id="radios-1" value="1" checked="checked">
+          1
```
[the remark this answers is not in the document]
Comment (jeffwang0516) [reply smart_campus/app/templates/app/station_edit.html]: it's for the radio button
like
O1 O2 O3...

File: smart_campus/app/static/app/js/station_new.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Lee-W) [reply smart_campus/app/static/app/js/station_new.js]: I've consulted one who is familiar with the FrontEnd.
He said it should be `for(var i=0; i < location_arr.length; i++){`.
Do you agree?

File: smart_campus/app/templates/app/station_edit.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (Lee-W) [reply smart_campus/app/templates/app/station_edit.html]: Could this number be generated by a loop?

File: smart_campus/app/views.py (hunk 9 above)
Comment (jeffwang0516) [reply smart_campus/app/views.py]: Should I return  HttpResponseForbidden for that error?

File: smart_campus/app/templates/app/station_edit.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeffwang0516) [reply smart_campus/app/templates/app/station_edit.html]: Yep, I'm doing that already for 2,3 4,..

File: smart_campus/app/static/app/js/station_new.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (jeffwang0516) [reply smart_campus/app/static/app/js/station_new.js]: Oh, I thought you're talking about the space in the comment
Agree with line 9

File: smart_campus/app/views.py (hunk 9 above)
Comment (jeffwang0516) [reply smart_campus/app/views.py]: I think that can(Permission.ADMIN) should go first
With Permission ADMIN can EDIT

File: smart_campus/app/forms.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class StationForm(forms.ModelForm):
+    beacon = forms.CharField()
+    lat = forms.FloatField()
+    lng = forms.FloatField()
+    img1 = forms.ImageField(required=False)
+    img2 = forms.ImageField(required=False)
+    img3 = forms.ImageField(required=False)
+    img4 = forms.ImageField(required=False)
```
[the remark this answers is not in the document]
Comment (jeffwang0516) [reply smart_campus/app/forms.py]: Found a way.
I'll rewrite it.

File: smart_campus/app/views.py (hunk 5 above)
Comment (jeffwang0516) [reply smart_campus/app/views.py]: I'll add a variable in the settings file.

File: smart_campus/app/views.py (hunk 9 above)
Comment (Lee-W) [reply smart_campus/app/views.py]: You should return HttpResponseForbidden at the very beginning if one is not equipped with proper permission.

Comment (Lee-W) [reply smart_campus/app/views.py]: Oh, I got your point.
I agree with that.

File: smart_campus/app/views.py (hunk 4 above)
Comment (jeffwang0516) [reply smart_campus/app/views.py]: yep
I'll add one in station, saving the url for the primary image

File: smart_campus/app/static/app/css/station_list.css [hunk 13]
```diff
@@ -0,0 +1,12 @@
+.btn-sq-lg{
+    width: 150px !important;
+    height: 150px !important;
+}
+.container{
+    margin-top: 100px;
+}
+.station-block{
+    font-family: 'Microsoft JhengHei';
```
[the remark this answers is not in the document]
Comment (jeffwang0516) [reply smart_campus/app/static/app/css/station_list.css]: I'll add other fonts.

File: smart_campus/app/management/commands/load_beacon_data.py [hunk 14]
```diff
@@ -0,0 +1,31 @@
+from django.core.management.base import BaseCommand, CommandError
+from django.contrib.gis.geos import Point
+
+import pandas as pd
```
Comment (Lee-W) [inline smart_campus/app/management/commands/load_beacon_data.py]: Make sure you add `pandas` to `requirements.txt`

File: smart_campus/app/management/commands/load_beacon_data.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+class Command(BaseCommand):
+    help = 'Insert beacon datas from xls file'
+
+    def add_arguments(self, parser):
+        parser.add_argument('path', nargs=1, type=str)
+
+    def handle(self, *args, **options):
+        try:
+            data = pd.read_excel(options['path'][0])
+            for i in range(0, len(data)):
```
Comment (Lee-W) [inline smart_campus/app/management/commands/load_beacon_data.py]: Use `for index, datum in enumerate(data)` instead.

File: smart_campus/app/templates/app/option_snippet.html [hunk 16]
```diff
@@ -0,0 +1,6 @@
+<!-- Template Snippet loading options -->
```
Comment (Lee-W) [inline smart_campus/app/templates/app/option_snippet.html]: To separate this kind of utility template with others, you can add a `_` prefix to the file name (i.e. `_option_snippet.html`).

File: smart_campus/app/templates/app/station_edit.html [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+{% block stylesheet %}
+  {{ block.super }}
+  <link rel="stylesheet" type="text/css" href="{% static "app/css/station_edit.css" %}" />
+{% endblock %}
+{% block content %}
+<div class="container content">
+  <!-- Display Error -->
+  {% for field in form %} 
+    {% for error in field.errors %}
+        <li class='text-danger'><strong>{{ field.label }}: {{ error|escape }}</strong></li>
```
Comment (Lee-W) [inline smart_campus/app/templates/app/station_edit.html]: Unify the style of using `{{}}`.
i.e. Do not mix use `{{ var }}` and `{{var}}`

File: smart_campus/app/views.py [hunk 18]
```diff
@@ -2,14 +2,21 @@
...
 from django.contrib import auth
-from django.http import HttpResponseRedirect, JsonResponse
+from django.http import HttpResponseRedirect, JsonResponse, HttpResponseForbidden
+from django.shortcuts import get_object_or_404
 from django.contrib.auth.decorators import login_required
 from django.contrib import messages
+from django.contrib.gis.geos import Point
+from django.conf import settings
+
+import os
```
Comment (Lee-W) [inline smart_campus/app/views.py]: I suggest that you can follow the import sequence of pep8.
i.e.`std libs` -> `3rd party libs` (including Django) -> `local packages`
But it's ok that you import Django first in a Django project.
Just note that you should follow the same rule across the project.

File: smart_campus/app/views.py [hunk 19]
```diff
@@ -113,23 +120,158 @@ def index(request):
...
+                # Clear old images
+                for img in StationImage.objects.filter(station=station):
+                    # Clear files on the disk
+                    # Delete model instance won't delete them
+                    if os.path.isfile(img.image.path):
+                        os.remove(img.image.path)
+                    img.delete()
+
+                # Add new images
+                for img_num in range(1, settings.MAX_IMGS_UPLOAD + 1):
```
Comment (Lee-W) [inline smart_campus/app/views.py]: Use `enumerate` instead.

File: smart_campus/app/views.py [hunk 20]
```diff
@@ -113,23 +120,158 @@ def index(request):
...
+                    img_name = 'img{0}'.format(img_num)
+                    if data[img_name]:
+                        StationImage.objects.create(
+                            station=station,
+                            image=data[img_name],
+                            is_primary=True if img_num == data['main_img_num'] else False
```
Comment (Lee-W) [inline smart_campus/app/views.py]: Write a statement here is kind of complicated.
I suggest you can save it as a variable.
e.g. 
```python
is_primary = ...
StationImage... (
    is_primary=is_primary
)
```

File: smart_campus/app/views.py [hunk 21]
```diff
@@ -113,23 +120,158 @@ def index(request):
...
+            station.owner_group = request.user.group
+            station.save()
+            # save first
+            # need an instance to add beacons
+            station.beacon_set.add(Beacon.objects.get(name=data['beacon']))
+            station.save()
+
+            # Add images
+            # data['imgx'] will be None if not uploaded
+            for img_num in range(1, settings.MAX_IMGS_UPLOAD + 1):
```
Comment (Lee-W) [inline smart_campus/app/views.py]: Use `enumerate` instead.

File: smart_campus/app/views.py [hunk 22]
```diff
@@ -113,23 +120,158 @@ def index(request):
...
+                img_name = 'img{0}'.format(img_num)
+                if data[img_name]:
+                    StationImage.objects.create(
+                        station=station,
+                        image=data[img_name],
+                        is_primary=True if img_num == data['main_img_num'] else False
+                    )
```
Comment (Lee-W) [inline smart_campus/app/views.py]: As mentioned above.

File: smart_campus/app/management/commands/load_beacon_data.py (hunk 15 above)
Comment (jeffwang0516) [reply smart_campus/app/management/commands/load_beacon_data.py]: enumerate(data) only shows column names
I'll use iterrows with pandas.DataFrame

File: smart_campus/app/templates/app/station_edit.html (hunk 17 above)
Comment (jeffwang0516) [reply smart_campus/app/templates/app/station_edit.html]: okay, 
checked in all templates

File: smart_campus/app/views.py (hunk 19 above)
Comment (jeffwang0516) [reply smart_campus/app/views.py]: Do you mean enumerate on the data?
images are not seperated from other form data

Comment (Lee-W) [reply smart_campus/app/views.py]: I don't get it.
What I mean is like the following code.

```python
for index, datum in enumerate(data, 1):
    is_primary = (img_num == data['main_img_num'])
    StationImage.objects.create(
        station=station,
        image=datum,
        is_primary=is_primary
    )
```

Also, I didn't notice the line `True if ... else ...`.
You need not do so.
`(img_num == data['main_img_num'])` works.
Even the brackets are not necessary.
They're for readability.

Comment (jeffwang0516) [reply smart_campus/app/views.py]: The **data** is a dict of datas from all the form field(include keys img1,img2,img3,..,imgN, main_img_num, beacon, category.......)

Comment (jeffwang0516) [reply smart_campus/app/views.py]: so it might not work?!

Comment (Lee-W) [reply smart_campus/app/views.py]: What about `for key, value in data.items()`?

Comment (jeffwang0516) [reply smart_campus/app/views.py]:
```
for key, value in data.items():
    if isinstance(value, django.core.files.uploadedfile.InMemoryUploadedFile):
        is_primary = (key == 'img{0}'.format(data['main_img_num']))
        StationImage.objects.create(
            station=station,
            image=value,
            is_primary=is_primary
        )
```

Comment (jeffwang0516) [reply smart_campus/app/views.py]: like this?!

Comment (jeffwang0516) [reply smart_campus/app/views.py]: Tested, this would work.

File: smart_campus/app/management/commands/load_beacon_data.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
+from django.core.management.base import BaseCommand, CommandError
+from django.contrib.gis.geos import Point
+
+import pandas as pd
+
+from app.models import Beacon, UserGroup
+
+
+class Command(BaseCommand):
+    help = 'Insert beacon datas from xls file'
```
Comment (Lee-W) [inline smart_campus/app/management/commands/load_beacon_data.py]: 'Insert beacon data from xls file'

File: smart_campus/app/management/commands/load_beacon_data.py [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+                        'name': datum['idname'],
+                        'description': datum['description'],
+                        'location': Point(y=float(datum['Latitude']), x=float(datum['Longitude'])),
+                        'owner_group': UserGroup.objects.filter(name=datum['OwnerGroup']).first()
+                    }
+                )
+        except FileNotFoundError as e:
+            raise CommandError(e)
+
+        self.stdout.write(self.style.SUCCESS('Successfully updated beacon datas'))
```
Comment (Lee-W) [inline smart_campus/app/management/commands/load_beacon_data.py]: 'Update beacon data succeeded.'

File: smart_campus/app/models.py [hunk 25]
```diff
@@ -227,12 +228,20 @@ def __str__(self):
...
     is_primary = models.BooleanField(default=False)
 
     def __repr__(self):
         return 'Image {img_id}'.format(img_id=self.id)
 
+    def save(self, *args, **kwargs):
+        super(StationImage, self).save(*args, **kwargs)
+
+        # auto save the url of the primary image to the station
+        if self.is_primary == True:
```
Comment (Lee-W) [inline smart_campus/app/models.py]: Typically, we do not use `==` to compare with `True` and `False`.
`self.is_primary is True` is more pythonic.
Is there any reason that you don't do so?

File: smart_campus/app/static/app/js/station_new.js [hunk 26]
```diff
@@ -28,23 +28,26 @@ function addMarker(title, lat, lng) {
 }
 
 var imgcnt=1;
-
+// Add new file upload field and eadio button
```
Comment (Lee-W) [inline smart_campus/app/static/app/js/station_new.js]: // Add new file upload field and radio button

File: smart_campus/app/models.py (hunk 25 above)
Comment (jeffwang0516) [reply smart_campus/app/models.py]: Uh no, I'll  fix this

**Pull request merged**